// @modelcontextprotocol/core/dist/index.cjs
const require_auth = require('./auth-CEFpHq32.cjs');

exports.AnnotationsSchema = require_auth.AnnotationsSchema;
exports.AudioContentSchema = require_auth.AudioContentSchema;
exports.BaseMetadataSchema = require_auth.BaseMetadataSchema;
exports.BlobResourceContentsSchema = require_auth.BlobResourceContentsSchema;
exports.BooleanSchemaSchema = require_auth.BooleanSchemaSchema;
exports.CallToolRequestParamsSchema = require_auth.CallToolRequestParamsSchema;
exports.CallToolRequestSchema = require_auth.CallToolRequestSchema;
exports.CallToolResultSchema = require_auth.CallToolResultSchema;
exports.CancelTaskRequestSchema = require_auth.CancelTaskRequestSchema;
exports.CancelTaskResultSchema = require_auth.CancelTaskResultSchema;
exports.CancelledNotificationParamsSchema = require_auth.CancelledNotificationParamsSchema;
exports.CancelledNotificationSchema = require_auth.CancelledNotificationSchema;
exports.ClientCapabilitiesSchema = require_auth.ClientCapabilitiesSchema;
exports.ClientNotificationSchema = require_auth.ClientNotificationSchema;
exports.ClientRequestSchema = require_auth.ClientRequestSchema;
exports.ClientResultSchema = require_auth.ClientResultSchema;
exports.CompatibilityCallToolResultSchema = require_auth.CompatibilityCallToolResultSchema;
exports.CompleteRequestParamsSchema = require_auth.CompleteRequestParamsSchema;
exports.CompleteRequestSchema = require_auth.CompleteRequestSchema;
exports.CompleteResultSchema = require_auth.CompleteResultSchema;
exports.ContentBlockSchema = require_auth.ContentBlockSchema;
exports.CreateMessageRequestParamsSchema = require_auth.CreateMessageRequestParamsSchema;
exports.CreateMessageRequestSchema = require_auth.CreateMessageRequestSchema;
exports.CreateMessageResultSchema = require_auth.CreateMessageResultSchema;
exports.CreateMessageResultWithToolsSchema = require_auth.CreateMessageResultWithToolsSchema;
exports.CreateTaskResultSchema = require_auth.CreateTaskResultSchema;
exports.CursorSchema = require_auth.CursorSchema;
exports.DiscoverRequestSchema = require_auth.DiscoverRequestSchema;
exports.DiscoverResultSchema = require_auth.DiscoverResultSchema;
exports.ElicitRequestFormParamsSchema = require_auth.ElicitRequestFormParamsSchema;
exports.ElicitRequestParamsSchema = require_auth.ElicitRequestParamsSchema;
exports.ElicitRequestSchema = require_auth.ElicitRequestSchema;
exports.ElicitRequestURLParamsSchema = require_auth.ElicitRequestURLParamsSchema;
exports.ElicitResultSchema = require_auth.ElicitResultSchema;
exports.ElicitationCompleteNotificationParamsSchema = require_auth.ElicitationCompleteNotificationParamsSchema;
exports.ElicitationCompleteNotificationSchema = require_auth.ElicitationCompleteNotificationSchema;
exports.EmbeddedResourceSchema = require_auth.EmbeddedResourceSchema;
exports.EmptyResultSchema = require_auth.EmptyResultSchema;
exports.EnumSchemaSchema = require_auth.EnumSchemaSchema;
exports.GetPromptRequestParamsSchema = require_auth.GetPromptRequestParamsSchema;
exports.GetPromptRequestSchema = require_auth.GetPromptRequestSchema;
exports.GetPromptResultSchema = require_auth.GetPromptResultSchema;
exports.GetTaskPayloadRequestSchema = require_auth.GetTaskPayloadRequestSchema;
exports.GetTaskPayloadResultSchema = require_auth.GetTaskPayloadResultSchema;
exports.GetTaskRequestSchema = require_auth.GetTaskRequestSchema;
exports.GetTaskResultSchema = require_auth.GetTaskResultSchema;
exports.IconSchema = require_auth.IconSchema;
exports.IconsSchema = require_auth.IconsSchema;
exports.IdJagTokenExchangeResponseSchema = require_auth.IdJagTokenExchangeResponseSchema;
exports.ImageContentSchema = require_auth.ImageContentSchema;
exports.ImplementationSchema = require_auth.ImplementationSchema;
exports.InitializeRequestParamsSchema = require_auth.InitializeRequestParamsSchema;
exports.InitializeRequestSchema = require_auth.InitializeRequestSchema;
exports.InitializeResultSchema = require_auth.InitializeResultSchema;
exports.InitializedNotificationSchema = require_auth.InitializedNotificationSchema;
exports.JSONArraySchema = require_auth.JSONArraySchema;
exports.JSONObjectSchema = require_auth.JSONObjectSchema;
exports.JSONRPCErrorResponseSchema = require_auth.JSONRPCErrorResponseSchema;
exports.JSONRPCMessageSchema = require_auth.JSONRPCMessageSchema;
exports.JSONRPCNotificationSchema = require_auth.JSONRPCNotificationSchema;
exports.JSONRPCRequestSchema = require_auth.JSONRPCRequestSchema;
exports.JSONRPCResponseSchema = require_auth.JSONRPCResponseSchema;
exports.JSONRPCResultResponseSchema = require_auth.JSONRPCResultResponseSchema;
exports.JSONValueSchema = require_auth.JSONValueSchema;
exports.LegacyTitledEnumSchemaSchema = require_auth.LegacyTitledEnumSchemaSchema;
exports.ListPromptsRequestSchema = require_auth.ListPromptsRequestSchema;
exports.ListPromptsResultSchema = require_auth.ListPromptsResultSchema;
exports.ListResourceTemplatesRequestSchema = require_auth.ListResourceTemplatesRequestSchema;
exports.ListResourceTemplatesResultSchema = require_auth.ListResourceTemplatesResultSchema;
exports.ListResourcesRequestSchema = require_auth.ListResourcesRequestSchema;
exports.ListResourcesResultSchema = require_auth.ListResourcesResultSchema;
exports.ListRootsRequestSchema = require_auth.ListRootsRequestSchema;
exports.ListRootsResultSchema = require_auth.ListRootsResultSchema;
exports.ListTasksRequestSchema = require_auth.ListTasksRequestSchema;
exports.ListTasksResultSchema = require_auth.ListTasksResultSchema;
exports.ListToolsRequestSchema = require_auth.ListToolsRequestSchema;
exports.ListToolsResultSchema = require_auth.ListToolsResultSchema;
exports.LoggingLevelSchema = require_auth.LoggingLevelSchema;
exports.LoggingMessageNotificationParamsSchema = require_auth.LoggingMessageNotificationParamsSchema;
exports.LoggingMessageNotificationSchema = require_auth.LoggingMessageNotificationSchema;
exports.ModelHintSchema = require_auth.ModelHintSchema;
exports.ModelPreferencesSchema = require_auth.ModelPreferencesSchema;
exports.MultiSelectEnumSchemaSchema = require_auth.MultiSelectEnumSchemaSchema;
exports.NotificationSchema = require_auth.NotificationSchema;
exports.NumberSchemaSchema = require_auth.NumberSchemaSchema;
exports.OAuthClientInformationFullSchema = require_auth.OAuthClientInformationFullSchema;
exports.OAuthClientInformationSchema = require_auth.OAuthClientInformationSchema;
exports.OAuthClientMetadataSchema = require_auth.OAuthClientMetadataSchema;
exports.OAuthClientRegistrationErrorSchema = require_auth.OAuthClientRegistrationErrorSchema;
exports.OAuthErrorResponseSchema = require_auth.OAuthErrorResponseSchema;
exports.OAuthMetadataSchema = require_auth.OAuthMetadataSchema;
exports.OAuthProtectedResourceMetadataSchema = require_auth.OAuthProtectedResourceMetadataSchema;
exports.OAuthTokenRevocationRequestSchema = require_auth.OAuthTokenRevocationRequestSchema;
exports.OAuthTokensSchema = require_auth.OAuthTokensSchema;
exports.OpenIdProviderDiscoveryMetadataSchema = require_auth.OpenIdProviderDiscoveryMetadataSchema;
exports.OpenIdProviderMetadataSchema = require_auth.OpenIdProviderMetadataSchema;
exports.PaginatedRequestParamsSchema = require_auth.PaginatedRequestParamsSchema;
exports.PaginatedRequestSchema = require_auth.PaginatedRequestSchema;
exports.PaginatedResultSchema = require_auth.PaginatedResultSchema;
exports.PingRequestSchema = require_auth.PingRequestSchema;
exports.PrimitiveSchemaDefinitionSchema = require_auth.PrimitiveSchemaDefinitionSchema;
exports.ProgressNotificationParamsSchema = require_auth.ProgressNotificationParamsSchema;
exports.ProgressNotificationSchema = require_auth.ProgressNotificationSchema;
exports.ProgressSchema = require_auth.ProgressSchema;
exports.ProgressTokenSchema = require_auth.ProgressTokenSchema;
exports.PromptArgumentSchema = require_auth.PromptArgumentSchema;
exports.PromptListChangedNotificationSchema = require_auth.PromptListChangedNotificationSchema;
exports.PromptMessageSchema = require_auth.PromptMessageSchema;
exports.PromptReferenceSchema = require_auth.PromptReferenceSchema;
exports.PromptSchema = require_auth.PromptSchema;
exports.ReadResourceRequestParamsSchema = require_auth.ReadResourceRequestParamsSchema;
exports.ReadResourceRequestSchema = require_auth.ReadResourceRequestSchema;
exports.ReadResourceResultSchema = require_auth.ReadResourceResultSchema;
exports.RelatedTaskMetadataSchema = require_auth.RelatedTaskMetadataSchema;
exports.RequestIdSchema = require_auth.RequestIdSchema;
exports.RequestMetaSchema = require_auth.RequestMetaSchema;
exports.RequestSchema = require_auth.RequestSchema;
exports.ResourceContentsSchema = require_auth.ResourceContentsSchema;
exports.ResourceLinkSchema = require_auth.ResourceLinkSchema;
exports.ResourceListChangedNotificationSchema = require_auth.ResourceListChangedNotificationSchema;
exports.ResourceRequestParamsSchema = require_auth.ResourceRequestParamsSchema;
exports.ResourceSchema = require_auth.ResourceSchema;
exports.ResourceTemplateReferenceSchema = require_auth.ResourceTemplateReferenceSchema;
exports.ResourceTemplateSchema = require_auth.ResourceTemplateSchema;
exports.ResourceUpdatedNotificationParamsSchema = require_auth.ResourceUpdatedNotificationParamsSchema;
exports.ResourceUpdatedNotificationSchema = require_auth.ResourceUpdatedNotificationSchema;
exports.ResultSchema = require_auth.ResultSchema;
exports.RoleSchema = require_auth.RoleSchema;
exports.RootSchema = require_auth.RootSchema;
exports.RootsListChangedNotificationSchema = require_auth.RootsListChangedNotificationSchema;
exports.SamplingContentSchema = require_auth.SamplingContentSchema;
exports.SamplingMessageContentBlockSchema = require_auth.SamplingMessageContentBlockSchema;
exports.SamplingMessageSchema = require_auth.SamplingMessageSchema;
exports.ServerCapabilitiesSchema = require_auth.ServerCapabilitiesSchema;
exports.ServerNotificationSchema = require_auth.ServerNotificationSchema;
exports.ServerRequestSchema = require_auth.ServerRequestSchema;
exports.ServerResultSchema = require_auth.ServerResultSchema;
exports.SetLevelRequestParamsSchema = require_auth.SetLevelRequestParamsSchema;
exports.SetLevelRequestSchema = require_auth.SetLevelRequestSchema;
exports.SingleSelectEnumSchemaSchema = require_auth.SingleSelectEnumSchemaSchema;
exports.StringSchemaSchema = require_auth.StringSchemaSchema;
exports.SubscribeRequestParamsSchema = require_auth.SubscribeRequestParamsSchema;
exports.SubscribeRequestSchema = require_auth.SubscribeRequestSchema;
exports.SubscriptionFilterSchema = require_auth.SubscriptionFilterSchema;
exports.SubscriptionsAcknowledgedNotificationParamsSchema = require_auth.SubscriptionsAcknowledgedNotificationParamsSchema;
exports.SubscriptionsAcknowledgedNotificationSchema = require_auth.SubscriptionsAcknowledgedNotificationSchema;
exports.SubscriptionsListenRequestParamsSchema = require_auth.SubscriptionsListenRequestParamsSchema;
exports.SubscriptionsListenRequestSchema = require_auth.SubscriptionsListenRequestSchema;
exports.SubscriptionsListenResultMetaSchema = require_auth.SubscriptionsListenResultMetaSchema;
exports.SubscriptionsListenResultSchema = require_auth.SubscriptionsListenResultSchema;
exports.TaskAugmentedRequestParamsSchema = require_auth.TaskAugmentedRequestParamsSchema;
exports.TaskCreationParamsSchema = require_auth.TaskCreationParamsSchema;
exports.TaskMetadataSchema = require_auth.TaskMetadataSchema;
exports.TaskSchema = require_auth.TaskSchema;
exports.TaskStatusNotificationParamsSchema = require_auth.TaskStatusNotificationParamsSchema;
exports.TaskStatusNotificationSchema = require_auth.TaskStatusNotificationSchema;
exports.TaskStatusSchema = require_auth.TaskStatusSchema;
exports.TextContentSchema = require_auth.TextContentSchema;
exports.TextResourceContentsSchema = require_auth.TextResourceContentsSchema;
exports.TitledMultiSelectEnumSchemaSchema = require_auth.TitledMultiSelectEnumSchemaSchema;
exports.TitledSingleSelectEnumSchemaSchema = require_auth.TitledSingleSelectEnumSchemaSchema;
exports.ToolAnnotationsSchema = require_auth.ToolAnnotationsSchema;
exports.ToolChoiceSchema = require_auth.ToolChoiceSchema;
exports.ToolExecutionSchema = require_auth.ToolExecutionSchema;
exports.ToolListChangedNotificationSchema = require_auth.ToolListChangedNotificationSchema;
exports.ToolResultContentSchema = require_auth.ToolResultContentSchema;
exports.ToolSchema = require_auth.ToolSchema;
exports.ToolUseContentSchema = require_auth.ToolUseContentSchema;
exports.UnsubscribeRequestParamsSchema = require_auth.UnsubscribeRequestParamsSchema;
exports.UnsubscribeRequestSchema = require_auth.UnsubscribeRequestSchema;
exports.UntitledMultiSelectEnumSchemaSchema = require_auth.UntitledMultiSelectEnumSchemaSchema;
exports.UntitledSingleSelectEnumSchemaSchema = require_auth.UntitledSingleSelectEnumSchemaSchema;
